pt/lessons/advanced/metaprogramming.md

```diff
@@ -0,0 +1,246 @@
+---
+layout: page
+title: Metaprogramação
+category: advanced
+order: 7
+lang: pt
+---
+
+Metaprogramação é o processo de utilização de código para escrever código. Em Elixir isso nos dá a capacidade de estender a linguagem para se adequar às nossas necessidades e dinamicamente alterar o código. Vamos começar observando como Elixir é representado sob o capô, em seguida, como modificá-lo e, finalmente, como podemos usar esse conhecimento para estendê-lo.
+
+
+Uma palavra de cautela: Metaprogramação é complicado e só deve ser usado quando for absolutamente necessário. O uso excessivo certamente levará a um código complexo que é difícil de entender e depurar.
+
+## Sumário
+
+- [Citação](#quote)
+- [Fechar Parênteses](#unquote)
+- [Macros](#macros)
+	- [Macros Privados](#macros-privados)
+	- [Higiene Macro](#macro-hygiene)
+	- [Obrigatório](#binding)
+
+## Citação
+
+O primeiro passo para metaprogramação é a compreensão de como as expressões são representados. Em Elixir da árvore abstrata de sintaxe (AST), a representação interna do nosso código, é composto de tuplas. Estas tuplas contêm três partes: o nome da função, metadados e argumentos da função.
+
+A fim de ver essas estruturas internas, Elixir fornece-nos com a função `quote/2`. Usando `quote/2` podemos converter o código Elixir em sua representação subjacente:
+
+```elixir
+iex> quote do: 42
+42
+iex> quote do: "Hello"
+"Hello"
+iex> quote do: :world
+:world
+iex> quote do: 1 + 2
+{:+, [context: Elixir, import: Kernel], [1, 2]}
+iex> quote do: if value, do: "True", else: "False"
+{:if, [context: Elixir, import: Kernel],
+ [{:value, [], Elixir}, [do: "True", else: "False"]]}
+iex(6)>
+```
+
+Observe os três primeiros não retornam tuplas? Há cinco literais que se voltar quando citou:
+
+```elixir
+iex> :atom
+:atom
+iex> "string"
+"string"
+iex> 1 # All numbers
+1
+iex> [1, 2] # Lists
+[1, 2]
+iex> {"hello", :world} # 2 element tuples
+{"hello", :world}
+```
+
+## Fechar Parênteses
```

Comment:
I think that it is not a good idea to translate `quote` and `unquote`. 

What do you think @pragmaticivan ?